### ISSUE TITLE
When NPCs deconstruct vehicles

### ISSUE THREAD

**Issue by Funguss** [CleverRaven/Cataclysm-DDA#35611]
1. NPCs do not respect attachment. An NPC can remove any part from a tile, be it a wheel hub without removing the wheel, or a frame and leaving all other parts attached.
2. NPCs can cause vehicles to have unattached areas, breaking their ability to deconstruct the vehicle.

# Steps To Reproduce

1. Find a vehicle.
2. Shift the NPC to one of the corners.
3. Set NPC to deconstruct via a zone and wait a few hours.
4. View resulting tiles, check for lack of frames or alternators without engines.
5. Wait until they suddenly decide they're (erroneously) finished and check for floaters. Occasionally the tile will be blank due to containing a part with no glyph (e.g. seatbelt).

# Expected behavior

NPCs should adhere to the rules governing deconstruction of vehicles.

# Versions and configuration

- OS: Windows
    - OS Version: MINGW/CYGWIN/MSYS2 on unknown Windows version
- Game Version: 0.D-9499-gd0db81d [64-bit]
- Graphics Version: Tiles
- Mods loaded: [
    Dark Days Ahead [dda],
    Aftershock [aftershock],
    C.R.I.T Expansion Mod [crt_expansion],
    Magiclysm [magiclysm],
    Makeshift Items Mod [makeshift],
    More Survival Tools [More_Survival_Tools],
    Mythological Replicas [nw_pack],
    Hydroponics [hydroponics],
    Mutant NPCs [mutant_npcs],
    No Rail Stations [No_Rail_Stations],
    Folding Parts pack [deoxymod],
    Vehicle Additions Pack [blazemod],
    Tanks and Other Vehicles [Tanks],
    Roadheader and other mining vehicles [Heavy miners],
    Disable NPC Needs [no_npc_food],
    No Fungal Monsters [No_Fungi],
    Beta National Guard Camp [national_guard_camp],
    More Locations [more_locations],
    Fuji's More Buildings [FujiStruct]
]

**Comment by sdp0et** [comment]
My NPCs will do the thing where they remove the frame first, but seem to have the opposite effect for wheels and hubs.  I have to manually removed wheels, otherwise they ignore that tile and deconstruct everything but the tiles the wheels are on and just enough frames between to not split the vehicle.

**Comment by Funguss** [comment]
Further testing, and I'm certain the bugs I've described come from having multiple NPCs working on the vehicle or doing some deconstruction yourself simultaneously.

I'm pretty sure the wheel thing is a separate bug as I'm yet to see an NPC remove wheels unless through the bug above, despite having every tool they need on their person. I can't see that I'm overlooking anything, but it's possible.

**Comment by davidpwbrown** [comment]
They *should* be respecting the same restrictions that the player does when removing parts...

But yes I think , as you say, its related to multiple workers on the same vehicle, which has proven to be a bit of a headache, I tried to cover that by saving the part indexes they are working on, and then all other workers check other workers stored part indexes and wont work on the same part, but obviously they are still getting in each others way, ill look into it.

**Comment by Funguss** [comment]
Yeah, good luck, it certainly seems like a headache! Shall I set up a separate issue for the wheel problem? I've discovered that NPCs definitely won't remove wheels for some reason, even the yoke and harness. I had forgotten they were classed as wheels (which is also a little odd, it means multiple must be laid out horizontally, not vertically, unless you have high mechanics. Not sure what I expect from that one, though, and how to get around it).
